src/Daemon/CrashWriter.cpp

```diff
@@ -0,0 +1,203 @@
+#include <Daemon/CrashWriter.h>
+
+#include <Poco/Util/Application.h>
+#include <Poco/Util/LayeredConfiguration.h>
+#include <Poco/Environment.h>
+
+#include <base/getMemoryAmount.h>
+#include <Common/logger_useful.h>
+#include <Common/Exception.h>
+#include <Common/ErrorCodes.h>
+#include <Common/SymbolIndex.h>
+#include <Common/StackTrace.h>
+#include <Common/getNumberOfCPUCoresToUse.h>
+#include <Core/ServerUUID.h>
+#include <IO/WriteHelpers.h>
+#include <IO/WriteBufferFromHTTP.h>
+
+#include <Common/config_version.h>
+
+using namespace DB;
+
+
+std::unique_ptr<CrashWriter> CrashWriter::instance;
+
+void CrashWriter::initialize(Poco::Util::LayeredConfiguration & config)
+{
+    instance.reset(new CrashWriter(config));
+}
+
+bool CrashWriter::initialized()
+{
+    return instance != nullptr;
+}
+
+CrashWriter::CrashWriter(Poco::Util::LayeredConfiguration & config)
+{
+    auto logger = getLogger("CrashWriter");
+
+    if (config.getBool("send_crash_reports.enabled", false))
+    {
+        endpoint = config.getString("send_crash_reports.endpoint");
+        server_data_path = config.getString("path", "");
+        LOG_DEBUG(logger, "Sending crash reports is initialized with {} endpoint (anonymized)", endpoint);
+    }
+    else
+    {
+        LOG_DEBUG(logger, "Sending crash reports is disabled");
+    }
+}
+
+void CrashWriter::onSignal(int sig, const std::string & error_message, const FramePointers & frame_pointers, size_t offset, size_t size)
+{
+    if (instance)
+        instance->sendError(Type::SIGNAL, sig, error_message, frame_pointers, offset, size);
+}
+
+void CrashWriter::onException(int code, const std::string & error_message, const FramePointers & frame_pointers, size_t offset, size_t size)
+{
+    if (instance)
+        instance->sendError(Type::EXCEPTION, code, error_message, frame_pointers, offset, size);
+}
+
+void CrashWriter::sendError(Type type, int sig_or_error, const std::string & error_message, const FramePointers & frame_pointers, size_t offset, size_t size)
+{
+    auto logger = getLogger("CrashWriter");
+    if (!instance)
+    {
+        LOG_INFO(logger, "Not sending crash report");
+        return;
+    }
+
+    try
+    {
+        LOG_INFO(logger, "Sending crash report");
```

Comment:
It would be great if we generated some kind of id and printed that in the fatal logs. It'd make it much easier to match bug reports / issues with received crashes